### DISCUSSION TITLE
Add proper readonly support to default Umbraco editors

### DISCUSSION THREAD

**Discussion by ronaldbarendse** [umbraco/Umbraco-CMS#12011] [Features and ideas]
Support for readonly editors in Umbraco is currently implemented in a very generic way by using `pointer-events: none;` CSS styling on the property in the backoffice UI. This is used to disable properties on non-default cultures when editing culture variant content nodes (when you see _Inherited from {default language}_ above the property label).

PR https://github.com/umbraco/Umbraco-CMS/pull/11730 ensures this is now also applied when you set a property to readonly using the `EditorModelEventManager` events/notifications.

Because it's still possible to move focus to editor inputs by clicking the label or tabbing into them, the current solution is merely a generic quick-fix. To properly support readonly editors, this needs to be implemented separately for each editor type. Some might just require adding `disabled="model.readonly"`, but more complex ones will probably require additional changes.

I've retrieved the list of default editors on a clean Umbraco 9.3 install, so this can be used to track the process of checking/implementing proper readonly support:

| Alias | Name | Supports readonly | Notes |
| - | - | - | - |
| Umbraco.BlockList | Block List | :x: |
| Umbraco.CheckBoxList | Checkbox list | :x: |
| Umbraco.ColorPicker | Color Picker | :x: |
| Umbraco.ContentPicker | Content Picker | :x: |
| Umbraco.DateTime | Date/Time | :x: |
| Umbraco.Decimal | Decimal | :x: |
| Umbraco.DropDown.Flexible | Dropdown | :x: |
| Umbraco.EmailAddress | Email address | :x: |
| Umbraco.ColorPicker.EyeDropper | Eye Dropper Color Picker | :x: |
| Umbraco.UploadField | File upload | :x: |
| UmbracoForms.FormPicker | Form Picker | :x: |
| UmbracoForms.ThemePicker | Forms Theme Picker | :x: |
| Umbraco.Grid | Grid layout | :x: |
| Umbraco.ImageCropper | Image Cropper | :x: |
| Umbraco.Label | Label | :x: |
| Umbraco.ListView | List view | :x: |
| Umbraco.MarkdownEditor | Markdown editor | :x: |
| Umbraco.MediaPicker3 | Media Picker | :x: |
| Umbraco.MediaPicker | Media Picker (legacy) | :x: |
| Umbraco.MemberGroupPicker | Member Group Picker | :x: |
| Umbraco.MemberPicker | Member Picker | :x: |
| Umbraco.MultiUrlPicker | Multi URL Picker | :x: |
| Umbraco.MultiNodeTreePicker | Multinode Treepicker | :x: |
| Umbraco.NestedContent | Nested Content | :x: |
| Umbraco.Integer | Numeric | :x: |
| Umbraco.RadioButtonList | Radio button list | :x: |
| Umbraco.MultipleTextstring | Repeatable textstrings | :x: |
| Umbraco.TinyMCE | Rich Text Editor | :x: |
| Umbraco.Slider | Slider | :x: |
| Umbraco.Tags | Tags | :x: |
| Umbraco.TextArea | Textarea | :x: |
| Umbraco.TextBox | Textbox | :x: |
| Umbraco.TrueFalse | Toggle | :heavy_check_mark: | Part of PR https://github.com/umbraco/Umbraco-CMS/pull/11587 (8.18.0 and 9.2.0) |
| Umbraco.UserPicker | User Picker | :x: |

**Comment by bjarnef** [comment]
Regarding the "readonly" toggle at _Inherited from {default language}_ it is actually possible to toggle this anyway via the property label:
https://github.com/umbraco/Umbraco-CMS/issues/9676

It would be useful if most property editors / components was able to configurate as readonly or disabled.

Sometimes we need to import data, where we would like to use a core property editor, which `admin` can manage, but not content editors. We have solved this either just using a label property, hiding the specific properties / group / tab hooking into `SendingContentModel` event or using a custom property editor. Some of the core components already have a `disabled` property.

Depending on the property editor `disabled` and `readonly` is not the same thing though, e.g. a text input with `readonly` attribute get focus, but with `disabled` attribute it isn't focused when navigating via keyboard.

**Reply by ronaldbarendse** [reply]
Correct, both https://github.com/umbraco/Umbraco-CMS/issues/9676 and https://github.com/umbraco/Umbraco-CMS/issues/9783 are the result of using the generic quick fix of disabling user input by only adding `pointer-events: none;` styling...

This still allows focusing/activating the editors in different ways, like labels that are bound to a form element or tabbing to it. We could add additional quick fixes for this by removing the label `for` attribute when it's readonly and adding `tabindex="-1"`, but that would indeed completely disable any user interaction, including selecting text from a readonly input 😞
